python/ray/data/aggregate.py

```diff
@@ -32,25 +44,42 @@ def __init__(
         for more details about accumulator-based aggregation.
 
         Args:
-            init: This is called once for each group
-                to return the empty accumulator.
+            init: This is called once for each group to return the empty accumulator.
                 For example, an empty accumulator for a sum would be 0.
-            accumulate: This is called once per row of the same group.
-                This combines the accumulator and the row,
-                returns the updated accumulator.
             merge: This may be called multiple times, each time to merge
                 two accumulators into one.
+            accumulate: This is called once per row of the same group.
+                This combines the accumulator and the row, returns the updated
+                accumulator.
+            vectorized_aggregate: This is used to calculate the aggregation for a single
+                block, and is vectorized alternative to accumulate. This will be given
+                the empty accumulator and the entire block, allowing for vectorized
+                aggregation of the block. This will only be used if
+                agg_fn.can_vectorize_for_block() returns True for the provided block.
             finalize: This is called once to compute the final aggregation
                 result from the fully merged accumulator.
             name: The name of the aggregation. This will be used as the output
                 column name in the case of Arrow dataset.
         """
+        if accumulate is None and vectorized_aggregate is None:
+            raise ValueError(
+                "At least one of accumulate or vectorized_aggregate must be provided."
+            )
         self.init = init
-        self.accumulate = accumulate
         self.merge = merge
+        self.accumulate = accumulate
+        self.vectorized_aggregate = vectorized_aggregate
         self.finalize = finalize
         self.name = name
 
+    def can_vectorize_for_block(self, block_acc: BlockAccessor[T]) -> bool:
```

Comment:
Do we need this? We can derive this from whether `self.acccumulate` and `self.vectorized_aggregate` are None? If only `self.accumulate` is not None, then this AggregateFn only supports single row mode. If only `self.vectorized_aggregate` is not None, then this AggregateFn only supports batch mode. If both are set, we prefer batch mode if possible.

Reply:
The reasoning for this was that desired behavior varies with the block type, since vectorized aggregation is currently not supported for simple blocks for the built-in aggregations given that it would reduce to the accumulator style + an extra data copy of each group slice view.

However, I'm about to push up a change that gets rid of this by taking the performance hit for the simple dataset case in order to make the API and implementation simpler, so this should be addressed.